### DISCUSSION TITLE
Implementing the von Mises model in Mfront

### DISCUSSION THREAD

**Discussion by kuateric** [thelfer/tfel#70] [Q&A] [answered]
Hi @thelfer 
I am going through the mfront learning curve. I took for that the easiest model that can be: the von Mises model. 
Based on the explanation in E.A. de Souza Neto, D. Peric, and D.R.J. Owen. Computational Methods for Plasticity: Theory and Applications. Wiley, 2011.
, I derived the closed form of the plastic multiplier: 
![image](https://user-images.githubusercontent.com/22998049/148281990-183cd3de-cf23-4469-b521-089ed7e8d514.png)

you did the same using a different method [here](http://tfel.sourceforge.net/IsotropicLinearHardeningPlasticity.html):
![image](https://user-images.githubusercontent.com/22998049/148282197-5da5ff59-ba97-4fbe-ab3c-78ce9679cbca.png)

Your approach is very condensed and quite difficult for me to understand right away. I will first stick to the Souza approach. For beginner, we can upload my version for dummys later on in your mfront website if you want. 

Now I would like to start the implementation. Prior to that i have some questions:
1. where do you derive the consistent tangent operator ? 
2. lets go through the integrator block: see comments
```c++
@Integrator{
  const auto lambda = computeLambda(young,nu);
  const auto mu     = computeMu(young,nu);
  eel += deto; // what is eel and deto ? 
  const auto se     = 2*mu*deviator(eel);
  const auto seq_e  = sigmaeq(se);
  const auto b      = seq_e-s0-H*p>stress{0};
  if(b){
    const auto iseq_e = 1/seq_e;
    const auto n      = eval(3*se/(2*seq_e));
    const auto cste   = 1/(H+3*mu);
    dp   = (seq_e-s0-H*p)*cste;
    eel -= dp*n;
    if(computeTangentOperator_){
      if(smt==CONSISTENTTANGENTOPERATOR){
    Dt = (lambda*Stensor4::IxI()+2*mu*Stensor4::Id()
          -4*mu*mu*(dp*iseq_e*(Stensor4::M()-(n^n))+cste*(n^n))); // this is the consistent tangent operator ?
      } else {
    Dt = lambda*Stensor4::IxI()+2*mu*Stensor4::Id();
      }
    }
  } else {
    if(computeTangentOperator_){
      Dt = lambda*Stensor4::IxI()+2*mu*Stensor4::Id();
    }
  }
  sig = lambda*trace(eel)*Stensor::Id()+2*mu*eel; // here you compute the stress for the next time increment right ? 
// I just see the elastic tensor in that equation 
// where does the consitent tangent operator intervene ?
}
```
3. is the code ready to be used in a solver ? I don't see the input data or the communication with fe-codes here

**Comment by thelfer** [comment]
Hi @kuateric,

Thanks for the feed-back. This example is indeed quite simple but the documentation is not complete. Furthermore, it is based on the `Default` DSL which has only a limited usage (far less than the `Implicit` DSL for example which is the object of most tutorials).

I updated the documentation to describe how is computed the consistent <https://thelfer.github.io/tfel/web/IsotropicLinearHardeningPlasticity.html>

**Reply by kuateric** [reply]
okay thanks, 
here is your version:
![image](https://user-images.githubusercontent.com/22998049/148371771-b79cf265-5b8e-4453-b72a-761c6d202669.png)
here is mine: 
![image](https://user-images.githubusercontent.com/22998049/148371943-9b3954d4-a9f6-4cb4-aeeb-cb3d9f025eaa.png)
with the following operation it is the same:
![image](https://user-images.githubusercontent.com/22998049/148374441-92574b0b-5c7d-4088-a28f-d67b6f692c25.png)

**Comment by kuateric** [comment]
did you see my comments here?
I can now identify the consistent tangent operator 
how do you derive the stress increment ?

> ```c++
> @Integrator{
>   const auto lambda = computeLambda(young,nu);
>   const auto mu     = computeMu(young,nu);
>   eel += deto; // what is eel and deto ? 
>   const auto se     = 2*mu*deviator(eel);
>   const auto seq_e  = sigmaeq(se);
>   const auto b      = seq_e-s0-H*p>stress{0};
>   if(b){
>     const auto iseq_e = 1/seq_e;
>     const auto n      = eval(3*se/(2*seq_e));
>     const auto cste   = 1/(H+3*mu);
>     dp   = (seq_e-s0-H*p)*cste;
>     eel -= dp*n;
>     if(computeTangentOperator_){
>       if(smt==CONSISTENTTANGENTOPERATOR){
>     Dt = (lambda*Stensor4::IxI()+2*mu*Stensor4::Id()
>           -4*mu*mu*(dp*iseq_e*(Stensor4::M()-(n^n))+cste*(n^n))); // this is the consistent tangent operator ?
>       } else {
>     Dt = lambda*Stensor4::IxI()+2*mu*Stensor4::Id();
>       }
>     }
>   } else {
>     if(computeTangentOperator_){
>       Dt = lambda*Stensor4::IxI()+2*mu*Stensor4::Id();
>     }
>   }
>   sig = lambda*trace(eel)*Stensor::Id()+2*mu*eel; // here you compute the stress for the next time increment right ? 
> // I just see the elastic tensor in that equation 
> // where does the consitent tangent operator intervene ?
> }
> ```

**Comment by thelfer** [comment, marked as answer]
Hi @kuateric,
I finished the documentation. You shall find the answers to your questions. Do not hesitate to make feed-backs and corrections.
Concerning the expression of the consistent tangent operators, I am pretty confident of my expressions (mostly because I checked the values of the consistent tangent operator by numerical differentiation in `MTest` :)).

**Comment by kuateric** [comment]
I could find the same expression of consistent tangent operator like you ;) 
![image](https://user-images.githubusercontent.com/22998049/148378433-d427e0b4-2c6f-437a-b963-d79ea210c3f2.png)

**Comment by kuateric** [comment]
Ahhh perfect, I see the documentation now. Thank you very much to find the time and the effort to write everything down for me.

**Reply by thelfer** [reply]
@kuateric, to be honest, I did not do it only for you :)

**Comment by kuateric** [comment]
![grafik](https://user-images.githubusercontent.com/22998049/148398931-0275ab1e-7f44-452c-a6f3-ed6cc75f102a.png)
![grafik](https://user-images.githubusercontent.com/22998049/148399953-67747d0c-d697-4d28-be2e-640fa0db881c.png)

**Comment by kuateric** [comment]
>We could also provide a tangent operator, but this would mean saving an auxiliary state variable stating if a plastic loading occured at the previous time step.

what do you mean exactly here ? I thought that we define the tangent operator in `@integrator` block. In which case does this cited tangent operator takes place?

**Reply by thelfer** [reply]
> what do you mean exactly here ? I thought that we define the tangent operator in `@integrator` block. In which case does this cited tangent operator takes place?

Just like there are various tangent operators (elastic, secant, tangent, consistent tangent), there may be several prediction operators (elastic, secant, tangent). Here we only provide the elastic one.

**Comment by kuateric** [comment]
Hi @thelfer 
what does the function `eval` exactly does :
`const auto n      = eval(3*se/(2*seq_e))`
knowing that the Prandl Reuss flow normal vector is:
![image](https://user-images.githubusercontent.com/22998049/148729368-58e7ec0c-dde9-45ab-9467-116f80c97b5e.png)

**Reply by thelfer** [reply]
Hi @kuateric,

`TFEL/Math` uses expression templates to perform lazy evaluations of most tensorial operations. See <https://thelfer.github.io/tfel/web/tfel-math.html#sec:tfel_math:expression_templates> for details.

The `eval` function evaluates an expression to get its results. I could have removed it here, but the normal is used multiple times, so I evaluate it once.

**Reply by kuateric** [reply]
thanks

**Comment by kuateric** [comment]
Why do you compute the elastic tangent operator twice in this block:
```c++
Dt = (lambda*Stensor4::IxI()+2*mu*Stensor4::Id() 
	      -4*mu*mu*(dp*iseq_e*(Stensor4::M()-(n^n))+cste*(n^n))); // elastoplastic tangent operator
      } else {
	Dt = lambda*Stensor4::IxI()+2*mu*Stensor4::Id(); // elastic tangent operator 
      }
    }
  } else {
    if(computeTangentOperator_){
      Dt = lambda*Stensor4::IxI()+2*mu*Stensor4::Id(); // elastic tangent operator again ? 
    }
```

**Reply by thelfer** [reply]
The new version of the `@Integrator`, which is consistent with the current documentation, is much clearer:

```cxx
@Integrator {
  const auto lambda = computeLambda(young, nu);
  const auto mu = computeMu(young, nu);
  // elastic loading
  eel += deto;
  if(computeTangentOperator_){
    Dt = lambda * Stensor4::IxI() + 2 * mu * Stensor4::Id();
  }
  // trial equivalent stress
  const auto se = 2 * mu * deviator(eel);
  const auto seq_e = sigmaeq(se);
  const auto plastic_loading = seq_e - s0 - H * p > stress{0};
  // plastic loading
  if (plastic_loading) {
    const auto iseq_e = 1 / seq_e;
    const auto n = eval(3 * se / (2 * seq_e));
    const auto cste = 1 / (H + 3 * mu);
    dp = (seq_e - s0 - H * p) * cste;
    eel -= dp * n;
    if ((computeTangentOperator_) && (smt == CONSISTENTTANGENTOPERATOR)) {
      Dt -= 4 * mu * mu *
            (dp * iseq_e * (Stensor4::M() - (n ^ n)) + cste * (n ^ n));
    }
  }
  // stress update
  sig = lambda * trace(eel) * Stensor::Id() + 2 * mu * eel;
}
```

**Comment by kuateric** [comment]
should the stress at the end of the time step not be instead:
`sig = lambda*trace(eel)*Stensor::Id()+2*mu*DEVIATOR(eel) `

**Reply by kuateric** [reply]
what about this ?

**Reply by thelfer** [reply]
No:
- `C++` is case-sensitive. So `DEVIATOR` and `deviator` are two different functions.
- If you want to use the deviator, you shall use the bulk compression moduli in the first part in place of the first Lamé coefficient `lambda`

**Comment by kuateric** [comment]
is my understanding of the eel computation right ? 

```c++
@Integrator {
  const auto lambda = computeLambda(young, nu);
  const auto mu = computeMu(young, nu);
  // elastic loading
  eel += deto; // strain_elastic = strain_total
  if(computeTangentOperator_){
    Dt = lambda * Stensor4::IxI() + 2 * mu * Stensor4::Id();
  }
  // trial equivalent stress
  const auto se = 2 * mu * deviator(eel);
  const auto seq_e = sigmaeq(se);
  const auto plastic_loading = seq_e - s0 - H * p > stress{0};
  // plastic loading
  if (plastic_loading) {
    const auto iseq_e = 1 / seq_e;
    const auto n = eval(3 * se / (2 * seq_e));
    const auto cste = 1 / (H + 3 * mu);
    dp = (seq_e - s0 - H * p) * cste;
    eel -= dp * n; // update of the elastic strain: strain_total - strain_plastic
    if ((computeTangentOperator_) && (smt == CONSISTENTTANGENTOPERATOR)) {
      Dt -= 4 * mu * mu *
            (dp * iseq_e * (Stensor4::M() - (n ^ n)) + cste * (n ^ n));
    }
  }
  // stress update
  sig = lambda * trace(eel) * Stensor::Id() + 2 * mu * eel;
}
```

**Reply by thelfer** [reply]
Nope :)

At the beginning of `@Integrator`, `eel` contains the value of the elastic strain at the beginning of the time step. 

> **About the plastic strain**
> 
> Note the implementation does not store the plastic strain. Its value at the beginning of the time step could have been retrieved using
> 
> ```cxx
> // eval is important here, as eel can be modified after this call
> const auto evp = eval(eto-eel);
> ```

The following line adds the strain increment `deto` to the elastic strain:

```cxx
  eel += deto;
```

This is the elastic solution, i.e. the solution if no plastic loading occurs. The elastic strain now have its value at the end of the time step if no plastic loading occurs.

**Reply by kuateric** [reply]
ok thanks

**Reply by thelfer** [reply]
Indeed, the code would have been clearer using increments. Then `eel` would not be modified inside the `@Integrator` code block and would keep its value at the beginning of the time step in this block.

This could be implemented as follows:

```cxx
@Integrator {
  const auto lambda = computeLambda(young, nu);
  const auto mu = computeMu(young, nu);
  // elastic loading
  deel = deto; 
  if(computeTangentOperator_){
    Dt = lambda * Stensor4::IxI() + 2 * mu * Stensor4::Id();
  }
  // trial equivalent stress
  const auto se = 2 * mu * deviator(eel + deel);
  const auto seq_e = sigmaeq(se);
  const auto plastic_loading = seq_e - s0 - H * p > stress{0};
  // plastic loading
  if (plastic_loading) {
    const auto iseq_e = 1 / seq_e;
    const auto n = eval(3 * se / (2 * seq_e));
    const auto cste = 1 / (H + 3 * mu);
    dp = (seq_e - s0 - H * p) * cste;
    deel -= dp * n;
    if ((computeTangentOperator_) && (smt == CONSISTENTTANGENTOPERATOR)) {
      Dt -= 4 * mu * mu *
            (dp * iseq_e * (Stensor4::M() - (n ^ n)) + cste * (n ^ n));
    }
  }
  // stress update
  sig = lambda * trace(eel + deel) * Stensor::Id() + 2 * mu * (eel + deel);
}
// note eel is then updated after @Integrator using its increment
```

However, this version is more cumbersome and less efficient, IMHO.

**Comment by kuateric** [comment]
what is the command to compile the mfront file again ?

**Reply by thelfer** [reply]
The basic command is:

```cxx
$ mfront --obuild --interface=generic your_file.mfront
```

**Reply by kuateric** [reply]
Can i compile two models in the same folder ? Because the folder `include` and `src` are created for each behaviour

**Reply by thelfer** [reply]
Yes, with some restrictions:

- It the behaviours have the same name (given by the `@Behaviour` keyword, this is **not** related to `MFront` file name), the second would overwrite the sources of the first one
- You shall not mix two different interfaces for the same behaviour

**Comment by kuateric** [comment]
how can i change the parameter in mtest ? 
this : 
# parameters
```
m.setBehaviour('generic', './src/libBehaviour.so', 'IsotropicLinearHardeningPlasticity')
m.setParameter('YoungModulus', 70.e9)
m.setParameter('PoissonRatio', 0.34)
m.setParameter('HardeningSlope', 10.e9)
m.setParameter('YieldStress', 300.e6)
```

leads to this:
`IsotropicLinearHardeningPlasticityParametersInitializer::set:  no parameter named 'YieldStress'
`

**Reply by thelfer** [reply]
This means that the first parameters are ok :)

You may find all the parameters associated with the parameters as follows:

```cxx
$ mfront-query --parameters IsotropicLinearHardeningPlasticity.mfront 
- YoungModulus (young): The Young modulus of an isotropic material
- PoissonRatio (nu): The Poisson ratio of an isotropic material
- HardeningSlope (H)
- YieldStrength (s0): la limite d'élasticité
- minimal_time_step_scaling_factor: minimal value for the time step scaling factor
- maximal_time_step_scaling_factor: maximal value for the time step scaling factor
```

In this case, the glossary name `YieldStress` is an alias to `YieldStrength`. A bit misleading, indeed.

**Reply by thelfer** [reply]
BTW, do no hesitate to open an issue for a better error message that would list available parameters.

**Reply by kuateric** [reply]
I don't see any improvement to do right now. 
I am just curious where the `YieldStrength` is defined in the code. I just find this:
`s0.setGlossaryName("YieldStress")`

**Reply by thelfer** [reply]
Glossary names may have aliases. So, even if you use YieldStress in the code, MFront will use YieldStrength as a canonical name among all aliases.

**Comment by kuateric** [comment]
the model is running.
![sheartest_strength](https://user-images.githubusercontent.com/22998049/148768250-ce0bd1e4-44db-4530-97db-6b4bba41dd53.png)

**Reply by thelfer** [reply]
Great news !

**Comment by kuateric** [comment]
Hi Thomas, 
I am trying to implement the implicit version of the von Mises model. Here is what I got so far: 
What do you think about it ? 

```cpp
@Parser Implicit;
@Behaviour vonMises_implicit;
@Author Eric Simo;
@Date   07/03/2022;
@Description{
  An implicit implementation of the von Mises model.

  The yield surface is defined by:
  "\["
  "  f(\sigmaeq,p) = \sigmaeq-s_{0}-H\,p"
  "\]"
}

// integration information
@Algorithm NewtonRaphson_NumericalJacobian;
@Brick StandardElasticity;
@Theta 1.0;
@Epsilon 1.e-14;
@ModellingHypotheses {".+"};
@RequireStiffnessTensor;

// state variables
@StateVariable StrainStensor eel;
eel.setGlossaryName("ElasticStrain");
@StateVariable strain p;
p.setGlossaryName("EquivalentPlasticStrain");

// material properties
@Parameter young =  70.e6;
young.setGlossaryName("YoungModulus");
@Parameter nu    =   0.34;
nu.setGlossaryName("PoissonRatio");
@Parameter H     =  10.e6;
H.setEntryName("HardeningSlope");
@Parameter s0    =  100.e6;
s0.setGlossaryName("YieldStress");

// local variables
@LocalVariable bool integration;

// elastic predictor and verification of the yield criterion
@InitLocalVariables{
  
  // elastic preditor
  const auto sig_el = computeElasticPrediction();

  // stress invariants
  const auto K = computeLambda(young, nu);
  const auto G = computeMu(young, nu);
  const auto s = 2 * G * deviator(eel); // trial deviatoric stress
  const auto q = sigmaeq(s); // elastic trial von Mises effective stress

  // von Mises yield function
  const auto f = q - s0 - H*p ;
  integration = false;
  if(f > 0.0){
    integration = true;
  }
}


// integration of constitutive equations
@Integrator{
  if(integration){
    
    // stress invariants
  const auto K = computeLambda(young, nu);
  const auto G = computeMu(young, nu);
  const auto J2 = 0.5*trace(square(s));
    
    // stress invariants first derivatives with respect to stresses 
    const auto d_J2_d_sig = s;

    // von Mises  yield function
    const auto f = q - s0 - H*p;

    // von Mises function derivatives 
    const auto d_f_d_J2 = sqrt(3)/(2*sqrt(J2));
    const auto d_f_d_sig = d_f_d_J2*d_J2_d_sig;

    // system of equations
    feel += dlambda*d_f_d_sig;
    flambda = f/D(0,0);
  }
}
```

**Reply by thelfer** [reply]
Hi Eric,

I have tons of comments :)

See below:

```cxx
@Parser Implicit;
@Behaviour vonMises_implicit;
@Author Eric Simo;
@Date 07 / 03 / 2022;
@Description {
  An implicit implementation of the von Mises model.

      The yield surface is defined by : "\["
                                        "  f(\sigmaeq,p) = \sigmaeq-s_{0}-H\,p"
                                        "\]"
}

// @UseQt true;

// integration information
@Algorithm NewtonRaphson_NumericalJacobian;
@Brick StandardElasticity;
@Theta 1.0;
@Epsilon 1.e-14;
// T.H. This may be important
@PerturbationValueForNumericalJacobianComputation 1.e-8;

@ModellingHypotheses{".+"};

// @RequireStiffnessTensor;
// T.H.: @RequireStiffnessTensor is not required here and inconsistent as you
// declare the Young modulus and Poisson ratio as parameters later.
//
// It is not required because you have an isotropic material and thus can have
// more efficient expressions using the elastic properties directly.
//
// Note the if you had not declared the Young modulus and Poisson ratio as
// parameters, the StandardElaticity brick would automatically declare them as
// material properties. Note also the importance of the glossary names that
// allows MFront to identify the Young modulus and Poisson ratio

// state variables

// @StateVariable StrainStensor eel;
// eel.setGlossaryName("ElasticStrain");

// T.H eel is automatically declared by the Implicit DSL.

@StateVariable strain p;
p.setGlossaryName("EquivalentPlasticStrain");

// material properties

// @Parameter young = 70.e6;
// young.setGlossaryName("YoungModulus");
// @Parameter nu = 0.34;
// nu.setGlossaryName("PoissonRatio");

// T.H. A better solution:
//
// @Parameter stress young = 70.e6;
// young.setGlossaryName("YoungModulus");
// @Parameter real nu = 0.34;
// nu.setGlossaryName("PoissonRatio");
//
// An even better solution:

@ElasticMaterialProperties{70e6, 0.34};

// @Parameter H = 10.e6;
// H.setEntryName("HardeningSlope");
// @Parameter s0 = 100.e6;
// s0.setGlossaryName("YieldStress");

// T.H. A better solution:
@Parameter stress H = 10.e6;
H.setEntryName("HardeningSlope");
@Parameter stress s0 = 100.e6;
s0.setGlossaryName("YieldStress");

// local variables
@LocalVariable bool integration;

// T.H. Let us compute the shear modulus only once
@LocalVariable stress G;

// elastic predictor and verification of the yield criterion
@InitLocalVariables {
  // T.H. shear modulus
  G = computeMu(young, nu);

  // elastic preditor
  const auto sig_el = computeElasticPrediction();

  // stress invariants

  // const auto s = 2 * G * deviator(eel); // trial deviatoric stress
  // T.H. This is wrong: you shall use the elatic prediction, not the
  // value of the deviatoric stress at the beginning of the time step

  // const auto q = sigmaeq(s); // elastic trial von Mises effective stress
  const auto q = sigmaeq(sig_el); // elastic trial von Mises effective stress

  // von Mises yield function
  const auto f = q - s0 - H * p;

  //   integration = false;
  //   if (f > 0.0) {
  //     integration = true;
  //   }

  // T.H. This is equivalent to:
  integration = f > stress{0};
}

// integration of constitutive equations
@Integrator {
  constexpr auto eeps = strain{1.e-14};
  const auto seps = eeps * young;

  // Little trick, this test avoids extra identation
  if (!integration) {
    return true;
  }

  // stress invariants
  //     const auto K = computeLambda(young, nu);
  //     const auto G = computeMu(young, nu);

  //    const auto J2 = 0.5 * trace(square(s));
  //
  // T.H. This is wrong:
  //
  // - s is undefined
  // - square retuns s * s (matrix multiplication)
  //
  // Note that J2 is sqrt((s|s)) which is not the von Mises stress

  const auto seq = sigmaeq(sig);

  //
  //     // stress invariants first derivatives with respect to stresses
  //     const auto d_J2_d_sig = s;

  // von Mises  yield function
  //    const auto f = seq - s0 - H * p;
  //
  // T.H. p is the value at the beginning of the time step, you must take the
  // estimation of its increment into account, as follows:
  const auto f = seq - s0 - H * (p + theta * dp);

  // normal
  const auto iseq = 1 / max(seq, seps); // avoid dividing by zero
  const auto df_dsig = 3 * deviator(sig) * (iseq / 2);
  //
  //   // von Mises function derivatives
  //   const auto d_f_d_J2 = sqrt(3) / (2 * sqrt(J2));
  //   const auto d_f_d_sig = d_f_d_J2 * d_J2_d_sig;

  // system of equations

  // feel += dlambda * d_f_d_sig;
  // T.H. : You must choose p or lambda as the variable !
  feel += dp * df_dsig;

  //  flambda = f / D(0, 0);
  //
  // T.H. Small changes as the stiffness tensor is no more defined
  fp = f / young;
}
```

Note that this example allowed to find a bug, which is cool: https://github.com/thelfer/tfel/issues/132

**Reply by kuateric** [reply]
thanks ;) It helps me a lot. I will go through it an let you know.

**Reply by thelfer** [reply]
my pleasure !

**Comment by kuateric** [comment]
I have now implement your improvements. I still have further comments:
```cpp
@Parser Implicit;
@Behaviour vonMises_implicit;
@Author Eric Simo;
@Date   07/03/2022;
@Description{
  An implicit implementation of the von Mises model.

  The yield surface is defined by:
  "\["
  "  f(\sigmaeq,p) = \sigmaeq-s_{0}-H\,p"
  "\]"
}

// integration information
@Algorithm NewtonRaphson_NumericalJacobian;
@Brick StandardElasticity;
@Theta 1.0;
@Epsilon 1.e-14;
@ModellingHypotheses {".+"};
@RequireStiffnessTensor;

// state variables
@StateVariable strain p;
p.setGlossaryName("EquivalentPlasticStrain");

// material properties
@ElasticMaterialProperties{70e6, 0.34};
@Parameter stress H     =  10.e6;
H.setEntryName("HardeningSlope");
@Parameter stress s0    =  100.e6;
s0.setGlossaryName("YieldStress");

// local variables
@LocalVariable bool integration;

// T.H. Let us compute the shear modulus only once
@LocalVariable stress G;

// elastic predictor and verification of the yield criterion
@InitLocalVariables{
  
  // elastic preditor
  const auto sig_el = computeElasticPrediction();

  // stress invariants
  const auto q = sigmaeq(sig_el); // elastic trial von Mises effective stress

   // von Mises yield function
  const auto f = q - s0 - H * p;
  integration = f > stress{0};
}

```

```cpp
// integration of constitutive equations
@Integrator{
  constexpr auto eeps = strain{1.e-14}; // these two line are used just to avoid 
  const auto seps=eeps * young; // dividing by zero in the normal expression ?
  if(!integration){
    return true; // if true, step is elastic ? 
  }
```
    
```cpp
const auto seq = sigmaeq(sig) ;
const auto f= seq - s0 - H * (p + theta * dp);  // this just because the equivalent plastic strain at time t +dt is equal to p + theta * dp? It is still not clear to me 
```
    
```cpp
// normal
const auto iseq = 1 / max(seq, seps); // avoid dividing by zero
const auto df_dsig = 3 * deviator(sig) * (iseq / 2); // this corresponds to the following expression ? 
```
![image](https://user-images.githubusercontent.com/22998049/158250565-a17820b5-07d7-4de9-be7b-73a850b17843.png)
![image](https://user-images.githubusercontent.com/22998049/158250598-abab9990-0cac-429c-a5f1-b214a068895f.png)


```cpp
// system of equations
feel += dp * df_dsig; // what eel by the way ? the total or the elastic strain. Does that corresponds to the expression below? 
fp = f / young; // I don't understand this expression  
}
```
![image](https://user-images.githubusercontent.com/22998049/158250937-abaec169-41c9-46b7-8c12-ccca06a1d7d5.png)

**Reply by thelfer** [reply]
@kuateric,

>   constexpr auto eeps = strain{1.e-14}; // these two line are used just to avoid 
>   const auto seps=eeps * young; // dividing by zero in the normal expression ?

Absolutely

>  if(!integration){
>    return true; // if true, step is elastic ? 
>  }

Yes (using the StandardElasticity brick) and no (in general) :)

The default initialisation of the jacobian and residual by the StandardElasticity brick:

- sets all the the residuals parts to the associated residual except the residual associated with the elastic strain which is set to deel-deto;
- sets the jacobian to identity

Returning true at the very beginning of `@Integrator` thus leads to a system with the following solution:

- deel = deto
- dp=0

which is the elastic solution.

> const auto f= seq - s0 - H * (p + theta * dp);  // this just because the equivalent plastic strain at time t +dt is equal to p + theta * dp? It is still not clear 

We evaluate the yield surface at t+theta*dp. The  StandardElasticity brick has already evaluated the stress sig at t+theta*dt using the elastic strain eel+theta*deel. The equivalent plastic strain at t+theta*dt is given by p+theta*dp.

See Equation "0" (the one before Equation 1) here: <https://thelfer.github.io/tfel/web/implicit-dsl.html>

Note that plasticity is a special case where we don't use the consistency condition, i.e. the expression of the equivalent plastic strain rate dp_dt but the physical condition stating that the system is exactly on the yield surface at t+theta*dt (and this is why we almost always take theta=1 when dealing with plasticity: this condition makes sense at the end of the time step). Hence the implicit equation associated with the equivalent plastic strain would naturally be 

f = 0. 

But, we want all our equations to be homogeneous to strain, so we take the equivalent form 

f/young = 0 

> const auto df_dsig = 3 * deviator(sig) * (iseq / 2); // this corresponds to the following expression ? 

No, this expression is general and comes from the definition of the normal as the derivative of the von Mises stress.
